docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+`compact_merkle_tree`. This can be done efficiently and only requires a few node accesses. As a
+crosscheck the root hash of the compact tree is compared against the current log root. If it does
+not match then log is corrupt and cannot be used.
+
+### Writing Leaves and Sequencing
+
+In the current RDBMS storage implementation log clients queue new leaves to the log, and the 
+`LeafData` record is created. Further writes are coordinated by a single sequencer, which adds 
+leaves to the tree. The sequencer is responsible for ordering the leaves and creating the 
+`SequencedLeafData` row linking the leaf and its sequence number. If there are duplicate leaves in
+the log then they share `LeafData` rows. Leaves that have not been sequenced are not accessible 
+via the log APIs.
+
+When leaves are added to the tree they are processed by a `merkle/compact_merkle_tree`, this causes a
+batched set of tree node updates to be applied. Each update is given its own revision number. The
+result is that a number of tree snapshots are directly available in storage. This contrasts with
+previous implementations where the tree is in RAM and only the most recent snapshot is directly
+available. Note that we may batch log updates so we don't necessarily have all intermediate tree
+snapshots directly available from storage.
+
+As an optimization intermediate nodes with only a left child are not stored. There is more detail
+on how this affects access to tree paths in the file `merkle/merkle_paths.go`. This differes from
+the previous C++ in-memory tree implementation. In summary the code must handle cases where there
+is no right sibling of the rightmost node in a level.
+
+Each batched write also produces an internal tree head, linking that write to the revision number.
+
+### Reading Log Nodes
+
+When nodes are accessed the typical pattern is to request the newest version of a node with a
+specified level and index that is not greater than a revision number that represents the tree at a
+specific size of interest.
+
+If there are multiple tree snapshots at the same tree size it does not matter which one is picked
+for this as they include the same set of leaves. The Log API provides support for determining the
+correct version to request when reading nodes.
+
+#### Reading Leaves
+
+API requests for leaf data involve a straightforward query by leaf data hash, leaf Merkle hash or
+leaf index followed by formatting and marshalling the data to be returned to the client.
+
+#### Serving Proofs
+
+API requests for proofs involve more work but both inclusion and consistency proofs follow the 
+same pattern.
+
+The node revision number for the tree size involved is obtained from storage and will be used in
+the subsequent node fetch.
+
+The tree path for the proof is calculated for a specified tree size using an algorithm based on the
+reference implementation of RFC 6962. The output of this is an ordered slice of `NodeIDs` that must 
```

Comment:
Is the slice of NodeIDs that specify a particular tree revision?  Is that the same revision as you determine 'for the tree size involved' (above)?  In that case was I wrong to suppose that every sub-tree has it's own separate revision number?